provider/kubernetes/weight_allocator.go

```diff
@@ -0,0 +1,163 @@
+package kubernetes
+
+import (
+	"fmt"
+
+	"github.com/containous/traefik/provider/label"
+	"gopkg.in/yaml.v2"
+	extensionsv1beta1 "k8s.io/api/extensions/v1beta1"
+)
+
+type weightAllocator interface {
+	getWeight(host, path, serviceName string) int
+}
+
+var _ weightAllocator = &defaultWeightAllocator{}
+var _ weightAllocator = &fractionalWeightAllocator{}
+
+type defaultWeightAllocator struct{}
+
+func (d *defaultWeightAllocator) getWeight(host, path, serviceName string) int {
+	return label.DefaultWeight
+}
+
+type ingressService struct {
+	host    string
+	path    string
+	service string
+}
+
+type fractionalWeightAllocator struct {
+	serviceWeights map[ingressService]int
+}
+
+func newFractionalWeightAllocator(ingress *extensionsv1beta1.Ingress, client Client) (*fractionalWeightAllocator, error) {
+	servicePercentageWeights, err := getServicesPercentageWeights(ingress)
+	if err != nil {
+		return nil, err
+	}
+
+	allocator := &fractionalWeightAllocator{
+		serviceWeights: map[ingressService]int{},
+	}
+
+	serviceInstanceCounts, err := getServiceInstanceCounts(ingress, client)
+	if err != nil {
+		return nil, err
+	}
+
+	serviceWeights := map[ingressService]int{}
+	for _, rule := range ingress.Spec.Rules {
+		// key: rule path string
+		// value: service names
+		fractionalPathServices := map[string][]string{}
+
+		// key: rule path string
+		// value: fractional percentage weight
+		fractionalPathWeights := map[string]percentageValue{}
+
+		for _, pa := range rule.HTTP.Paths {
+			if _, ok := fractionalPathWeights[pa.Path]; !ok {
+				fractionalPathWeights[pa.Path] = newPercentageValueFromFloat64(1)
+			}
+
+			if weight, ok := servicePercentageWeights[pa.Backend.ServiceName]; ok {
+				ingSvc := ingressService{
+					host:    rule.Host,
+					path:    pa.Path,
+					service: pa.Backend.ServiceName,
+				}
+
+				serviceWeights[ingSvc] = weight.computeWeight(serviceInstanceCounts[ingSvc])
+
+				fractionalPathWeights[pa.Path] = fractionalPathWeights[pa.Path].sub(weight)
+				if fractionalPathWeights[pa.Path].toFloat64() < 0 {
+					return nil, fmt.Errorf("percentage value %s out of range", fractionalPathWeights[pa.Path].String())
+				}
+			} else {
+				fractionalPathServices[pa.Path] = append(fractionalPathServices[pa.Path], pa.Backend.ServiceName)
+			}
+		}
+
+		for pa, svcs := range fractionalPathServices {
+			totalFractionalInstanceCount := 0
+
+			for _, svc := range svcs {
+				totalFractionalInstanceCount += serviceInstanceCounts[ingressService{
+					host:    rule.Host,
+					path:    pa,
+					service: svc,
+				}]
+			}
```

Comment:
`totalFractionalInstanceCount` is used, seems like something is missing.

Reply:
@ldez Done